frontend/src/pages/projects/screens/spawner/SpawnerPage.tsx

```diff
@@ -99,8 +101,13 @@ const SpawnerPage: React.FC<SpawnerPageProps> = ({ existingNotebook }) => {
   React.useEffect(() => {
     if (notebookSize) {
       setSelectedSize(notebookSize.name);
+    } else {
+      if (existingNotebook && !sizes.some((size) => size === customSize)) {
+        sizes.unshift(customSize);
+      }
```

Comment:
For this in the Edit screen:
<img width="677" alt="Screenshot 2024-07-09 at 11 03 21 PM" src="https://github.com/opendatahub-io/odh-dashboard/assets/22885912/d21470d2-5085-45f8-a4e4-d396a00be434">

We can't select a size that's not in sizes.

Reply:
You shouldn't mutate a variable you didn't create like this. `unshift` mutates the array, which could have negative impacts on React's ability to diff between renders -- in this case, you're just mutating another runtime variable, but it could have side-effects on other functions run before or after your code. In React, you should be having a single source of truth for what are the available sizes and getting that updated with the `Keep custom size` and not trying to adhoc change the value later.

So maybe we need to rework this effort so it's more aligned... maybe something like this?

Can you take all of:
```
  const { size: notebookSize } = useNotebookDeploymentSize(existingNotebook);
  React.useEffect(() => {
    if (notebookSize) {
      setSelectedSize(notebookSize.name);
    } else {
      if (existingNotebook && !sizes.some((size) => size === CUSTOM_NOTEBOOK_SIZE)) {
        sizes.unshift(CUSTOM_NOTEBOOK_SIZE);
      }
      setSelectedSize(CUSTOM_NOTEBOOK_SIZE.name);
    }
  }, [notebookSize, setSelectedSize, sizes, existingNotebook]);
```

And create a new custom hook that uses this:
```
const { selectedSize, setSelectedSize, sizes } = useNotebookSize();
```

And produces a set of sizes & selectedSize to be used by the form, which manages both the Edit with custom, the create, and the edit with a known size.

Reply:
Added a custom hook and this custom hook works better.